### ISSUE TITLE
Anonymous FTP server error

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4842]
|<img alt="sanxiyn's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| sanxiyn reported|
|-|-|
|Trac ID|trac#4842|
|Type|defect|
|Created|2011-02-01 08:11:56Z|

How to reproduce:

1. Run Twisted's FTP server with anonymous access. (I used doc/core/examples/ftpserver.py.)

2. Connect to the server as anonymous and delete a file.

Expected: FTP error.  Instead:

```
Traceback (most recent call last):
  File "/System/Library/Frameworks/Python.framework/Versions/2.6/Extras/lib/python/twisted/protocols/policies.py", line 72, in dataReceived
    self.wrappedProtocol.dataReceived(data)
  File "/System/Library/Frameworks/Python.framework/Versions/2.6/Extras/lib/python/twisted/protocols/basic.py", line 231, in dataReceived
    why = self.lineReceived(line)
  File "/System/Library/Frameworks/Python.framework/Versions/2.6/Extras/lib/python/twisted/protocols/ftp.py", line 700, in lineReceived
    d.addCallbacks(processSucceeded, processFailed)
  File "/System/Library/Frameworks/Python.framework/Versions/2.6/Extras/lib/python/twisted/internet/defer.py", line 186, in addCallbacks
    self._runCallbacks()
--- <exception caught here> ---
  File "/System/Library/Frameworks/Python.framework/Versions/2.6/Extras/lib/python/twisted/internet/defer.py", line 328, in _runCallbacks
    self.result = callback(self.result, *args, **kw)
  File "/System/Library/Frameworks/Python.framework/Versions/2.6/Extras/lib/python/twisted/protocols/ftp.py", line 673, in processFailed
    self.sendLine(err.value.response())
  File "/System/Library/Frameworks/Python.framework/Versions/2.6/Extras/lib/python/twisted/protocols/ftp.py", line 237, in response
    return RESPONSE[self.errorCode] % self.errorMessage
exceptions.TypeError: not all arguments converted during string formatting
```

Attachments:

* [anonuserdeniederror.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/7fc/7fce3e1c20139ed4f6013a7a3028744aad2479dd/5d7b2c7b87cc3a1944ab844ec517c902e6de3c8e.patch) (1097 bytes) - added by eternicode on 2011-02-02 23:18:33Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__4842 4842
type__defect defect
reporter__sanxiyn sanxiyn
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__ftp ftp
keywords__ 
time__1296547916000000 1296547916000000
changetime__1296714889000000 1296714889000000
version__None None
owner__ 
cc__itamarst cc__eternicode
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="eternicode's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|eternicode commented|
|-|-|

See the definitions for [FTPCmdError](http://twistedmatrix.com/trac/browser/trunk/twisted/protocols/ftp.py#L225) and [AnonUserDeniedError](http://twistedmatrix.com/trac/browser/trunk/twisted/protocols/ftp.py#L250).

The ANON_USER_DENIED (500.3) response doesn't have any formatting placeholders in it, and thus nothing to interpolate (None,) to.  The simple solution is to remove the None arg from the ```FTPCmdError.__init__``` call, but it probably needs some tests to go along with it.

**Comment by twisted-trac** [comment]
|<img alt="eternicode's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|eternicode commented|
|-|-|

Attached a patch that fixes the ```__init__``` and adds a unit test for anonymous ftp write methods.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_6"></a>|@glyph set owner to @glyph|
|-|-|

Thanks very much for your patch!  I'll apply a slightly edited version of it.  A couple of minor notes, which you will see I've adjusted in the commit:

  - You need to add a NEWS file for every change, so it's helpful to attach one to patches.  In this case it's a minor bugfix that doesn't merit a mention, so I added an empty '.misc' file just to indicate that it's been fixed.  [More details on this requirement are here.](http://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles)
  - The modern standard for unit tests involves names like `test_fooBarBaz` rather than `testFooBarBaz`.  `test_ftp` is an old, bad example, so I can understand why you might not have noticed.
  - Tests should always have docstrings.  The style in the commit should generally be followed.
  - If you can, in the future, try to write unit tests which will _fail_ quickly when the code is broken, rather than generating random errors that don't trigger their assertions, or hanging until the test timeout.  Again, `test_ftp`'s harness doesn't set the best example, so I'm going to commit your test as-is (from the logged traceback, it's pretty clear that something's wrong).

Again, thanks for the patch, and I'm glad to kill yet another FTP bug :).  I'll be committing it shortly.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_7"></a>|@glyph set status to `closed`|
|-|-|

(In [30601]) Fix a confusing traceback / hang when anonymous users attempt a write operation
via FTP; replace it with an appropriate protocol error.

Author: eternicode

Reviewer: glyph

Fixes [#4842](https://github.com/twisted/twisted/issues/4842)

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
